### PR TITLE
[InstCombine] Add transforms for `icmp uPred (trunc x),(truncOrZext(y))` -> `icmp uPred x,y`

### PR DESCRIPTION
**Pull request by goldsteinn**
- [InstCombine] Add tests for transforming `(icmp eq/ne trunc(x), truncOrZext(y))`; NFC
- [InstCombine] Add transforms for `(icmp uPred (trunc x),(truncOrZext(y)))`->`(icmp uPred x,y)`

### CODE REVIEW

**Comment by nikic** [comment]
Should this be guarded by a shouldChangeType() heuristic? Say you're doing `icmp eq (trunc i256 to i32, trunc i256 to i32)`. Is folding that to `icmp eq i256` a good idea?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1553]:
```suggestion
  if (Cmp.isSigned())
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1592]:
```suggestion
  return new ICmpInst(Pred, X, NewY);
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1572]:
```suggestion
  else if (match(&Cmp, m_c_ICmp(Pred, m_Trunc(m_Value(X)),
                                m_OneUse(m_ZExt(m_Value(Y))))))
    YIsZext = true;
  else
    return nullptr;
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1580]:
```suggestion
  KnownBits KnownX = computeKnownBits(X, Q.DL, /*Depth*/ 0, Q.AC, Q.CxtI, Q.DT);
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1581]:
```suggestion
  if (KnownX.countMinLeadingZeros() < KnownX.getBitWidth() - TruncBits)
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1580]: I believe we have a public overload that directly takes SimplifyQuery nowadays.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1572]: Generally if any statement has braces I do for all, but no strong feelings about this.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1553]: Agreed less code although think the original shows intents of the lemma better. But no strong feelings. Changing.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1582]:
```suggestion
  if (!shouldChangeType(TruncTy, X->getType()))
```
I don't think this transform is safe for vectors (e.g., `<16 x i32>` --> `<16 x i64>`).
`shouldChangeType` looks safer because it doesn't support vectors.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1582]: Why isn't it safe for vectors?
https://alive2.llvm.org/ce/z/RTgGbb

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1582]: Or you mean it may regress on something like x86 b.c of the `sext` that typically follows to ele width?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1582]: It's the latter I'm concerned about.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1582]: Doesn't seem to get worse codegen: https://godbolt.org/z/naE651zPe

**Review by dtcxzyw — APPROVED** [review]
LGTM. Waiting for additional approval from @nikic.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1566]: Missing `getSwappedPredicate()`?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1587]:
```suggestion
  if (KnownX.countMaxActiveBits() > TruncBits)
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1593]:
```suggestion
    if (KnownY.countMaxActiveBits() > TruncBits)
```

**Review by nikic — COMMENTED** [review]
Can you please check whether this has compile-time impact (due to computeKnownBits)?

(We could reuse the KnownBits from foldICmpUsingKnownBits if necessary.)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1533]: Fixed + added assosiated tested.

**Comment by goldsteinn** [comment]
> Can you please check whether this has compile-time impact (due to computeKnownBits)?
> 
> (We could reuse the KnownBits from foldICmpUsingKnownBits if necessary.)

Seems like minimal affect: https://llvm-compile-time-tracker.com/compare.php?from=2b71f91b06ad4f5a0c54725b06283fd731620b92&to=2343dfccca46d158de240fca8e0588ae4cc46203&stat=instructions:u

Slight regression in ThinLTO. Larger improvement in stage2 + O3.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**
